metals/src/main/scala/scala/meta/internal/metals/debug/DebugDiscovery.scala

```diff
@@ -162,9 +172,226 @@ class DebugDiscovery(
+  }
+
+  private def findMainClasses(
+      textDocument: TextDocument,
+      classes: BuildTargetClasses.Classes,
+  ): Seq[b.ScalaMainClass] = {
+    val mains = for {
+      occurrence <- textDocument.occurrences
+      if occurrence.role.isDefinition || occurrence.symbol == "scala/main#"
+      symbol = occurrence.symbol
+      mainClass <- {
+        val normal = classes.mainClasses.get(symbol)
+        val fromAnnot = DebugDiscovery
+          .mainFromAnnotation(occurrence, textDocument)
+          .flatMap(classes.mainClasses.get(_))
+        List(normal, fromAnnot).flatten
+      }
+    } yield mainClass
+
+    if (mains.nonEmpty) mains
+    else DebugDiscovery.syntheticMains(textDocument, classes)
+  }
+
+  private def discoverTestCases(
+      path: AbsolutePath
+  ): List[(String, List[TestCaseEntry])] = {
+    import scala.jdk.CollectionConverters._
+    import scala.meta.internal.metals.testProvider.TestExplorerEvent.AddTestCases
+
+    val testUpdates = testProvider.discoverTests(Some(path))
+    testUpdates
+      .flatMap(_.events.asScala)
+      .collect { case AddTestCases(fqn, _, testCases) =>
+        (fqn, testCases.asScala.toList)
+      }
+      .toList
+  }
+
+  private def calculateMainDistances(
+      textDocument: TextDocument,
+      mains: Seq[b.ScalaMainClass],
+      classes: BuildTargetClasses.Classes,
+      position: Position,
+  ): Seq[(Int, Int, String, String, Any)] = {
+    mains.flatMap { mainClass =>
+      val occOpt = textDocument.occurrences.find { occ =>
+        occ.role.isDefinition && (
+          classes.mainClasses.get(occ.symbol).contains(mainClass) ||
+            DebugDiscovery
+              .mainFromAnnotation(occ, textDocument)
+              .flatMap(classes.mainClasses.get(_))
+              .contains(mainClass)
+        )
+      }
+
+      val syntheticOccOpt = if (occOpt.isEmpty) {
+        textDocument.symbols
+          .find(_.symbol.contains(mainClass.getClassName.split('.').last))
+          .flatMap(symbolInfo =>
+            textDocument.occurrences.find(_.symbol == symbolInfo.symbol)
+          )
+      } else None
+
+      (occOpt.orElse(syntheticOccOpt)).flatMap(_.range).map { range =>
+        val lineDiff = math.abs(range.startLine - position.getLine)
+        val charDiff = math.abs(range.startCharacter - position.getCharacter)
+        (lineDiff, charDiff, "main", mainClass.getClassName, mainClass)
+      }
+    }
+  }
+
+  private def calculateTestCaseDistances(
+      allTestCases: List[(String, List[TestCaseEntry])],
+      position: Position,
+  ): Seq[(Int, Int, String, String, String)] = {
+    allTestCases.flatMap { case (suiteFqn, cases) =>
+      cases.map { entry =>
+        val range = entry.location.getRange()
+        val start = range.getStart()
+        val lineDiff = math.abs(start.getLine() - position.getLine)
+        val charDiff = math.abs(start.getCharacter() - position.getCharacter)
+        (lineDiff, charDiff, "testCase", entry.name, suiteFqn)
```

Comment:
Wouldn't lineDiff be enough? I don't think we ever have multiple tests or main at the same line